Define functions on scalars and vectorise them

ChristianZimpelmann opened the pull request:
### What problem do you want to solve?

- Make functions easier to read and understand
- Make code much faster

### Todo

- [x] Implement functions to aggregate values on hh or tu level
- [x] Rewrite all functions such that they are defined on scalars
- [x] Implement vectorization (for numpy ~~, numba,~~ and jax)
- [x] Rewrite `interface.py` such that all columns are on individual level
- [x] If it is a relevant contribution, put it in `CHANGES.rst`.

Closes #281

hmgaudecker commented:
Tests still fail due to:

```pytb
gettsim/aggregation_numpy.py:56: in grouped_min
    out_on_hh = npg.aggregate(group_id, column, func="min", fill_value=fill_value)
/home/hmg/miniconda3/envs/gettsim/lib/python3.9/site-packages/numpy_groupies/aggregate_numpy.py:291: in aggregate
    return _aggregate_base(group_idx, a, size=size, fill_value=fill_value,
/home/hmg/miniconda3/envs/gettsim/lib/python3.9/site-packages/numpy_groupies/aggregate_numpy.py:280: in _aggregate_base
    ret = func(group_idx, a, flat_size, fill_value=fill_value, dtype=dtype,
/home/hmg/miniconda3/envs/gettsim/lib/python3.9/site-packages/numpy_groupies/aggregate_numpy.py:82: in _min
    else np.finfo(a.dtype).max
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'numpy.finfo'>, dtype = <class 'numpy.datetime64'>

    def __new__(cls, dtype):
        try:
            dtype = numeric.dtype(dtype)
        except TypeError:
            # In case a float instance was given
            dtype = numeric.dtype(type(dtype))
    
        obj = cls._finfo_cache.get(dtype, None)
        if obj is not None:
            return obj
        dtypes = [dtype]
        newdtype = numeric.obj2sctype(dtype)
        if newdtype is not dtype:
            dtypes.append(newdtype)
            dtype = newdtype
        if not issubclass(dtype, numeric.inexact):
>           raise ValueError("data type %r not inexact" % (dtype))
E           ValueError: data type <class 'numpy.datetime64'> not inexact

/home/hmg/miniconda3/envs/gettsim/lib/python3.9/site-packages/numpy/core/getlimits.py:473: ValueError
```

Should I just ignore? Should we skip those?

ChristianZimpelmann commented:
The PR is now ready for review. A few comments:

- I haven't implemented user-provided aggregation functions, yet. Will do so tomorrow.
- Also missing is a final update of the tutorials.
- I added a `_hh` suffix to a few column names which are expected to be constant within households.
- There is now a separate test workflow that runs jax and uses the environment `environment_linux_jax.yml`. While adding those tests, I dropped support for python 3.7.

hmgaudecker commented on gettsim/social_insurance_contributions/beitr_bemess_grenzen.py:
```diff
@@ -45,20 +39,19 @@ def _ges_krankenv_beitr_bemess_grenze_m(
-    return out.astype(float)
+    params = soz_vers_beitr_params["beitr_bemess_grenze_m"]["ges_krankenv"]
+
+    if wohnort_ost:
```
```suggestion
    out = params["ost"] if wohnort_ost else params["west"]
```
is much more readable (that's why the construct was created fairly recently). No need to change everywhere, but this particular instance appears so often that it seems worth a search & replace to me.

hmgaudecker commented on gettsim/social_insurance_contributions/eink_grenzen.py:
```diff
@@ -15,23 +11,16 @@ def mini_job_grenze(wohnort_ost: BoolSeries, soz_vers_beitr_params: dict):
-            ],
-        }
-    )
+    mini_job_params = soz_vers_beitr_params["geringfügige_eink_grenzen_m"]["mini_job"]
```
```suggestion
    params = soz_vers_beitr_params["geringfügige_eink_grenzen_m"]["mini_job"]
```
As above. I like this shortening, but let's be consistent with it.

hmgaudecker commented on gettsim/social_insurance_contributions/eink_grenzen.py:
```diff
@@ -76,16 +63,17 @@ def in_gleitzone(
     Returns
     -------
-    BoolSeries indicating individual's income is in midi-job range.
+    Boolean indicating individual's income is in midi-job range.
```
```suggestion
    Whether individual's income is in midi-job range.
```

hmgaudecker commented on gettsim/social_insurance_contributions/eink_grenzen.py:
```diff
@@ -48,16 +37,14 @@ def geringfügig_beschäftigt(
     Returns
     -------
-    BoolSeries indicating if person earns less than marginal employment threshold.
+    Boolean indicating if person earns less than marginal employment threshold.
```
```suggestion
    Whether person earns less than marginal employment threshold.
```

hmgaudecker commented on gettsim/social_insurance_contributions/eink_grenzen.py:
```diff
@@ -76,16 +63,17 @@ def in_gleitzone(
-) -> FloatSeries:
+    bruttolohn_m: float, soz_vers_beitr_params: dict,
+) -> float:
     """Select income subject to social insurance contributions for midi job.
```
```suggestion
    """Income subject to social insurance contributions for midi job.
```

hmgaudecker commented on gettsim/social_insurance_contributions/eink_grenzen.py:
```diff
@@ -98,15 +86,13 @@ def midi_job_bemessungsentgelt_m(
     Returns
     -------
-    FloatSeries with the income subject to social insurance contributions for midi job.
+    float with the income subject to social insurance contributions for midi job.
```
Do we really need to repeat the dtype?

hmgaudecker commented on gettsim/social_insurance_contributions/ges_pflegev.py:
```diff
@@ -1,13 +1,6 @@
+    hat_kinder: bool, alter: int, soz_vers_beitr_params: dict,
+) -> bool:
     """
     Create boolean Series indicating addtional care insurance contribution for
```
```suggestion
    Whether additional care insurance contribution for childless individuals applies.
```

hmgaudecker commented on gettsim/social_insurance_contributions/ges_pflegev.py:
```diff
@@ -1,13 +1,6 @@
+) -> bool:
     """
     Create boolean Series indicating addtional care insurance contribution for
     childless individuals.
```
```suggestion
```

hmgaudecker commented on gettsim/social_insurance_contributions/ges_pflegev.py:
```diff
@@ -199,14 +198,15 @@ def _ges_pflegev_beitr_midi_job_m_m(
-    )
+    # Add additional contribution for childless individuals
+    if ges_pflegev_zusatz_kinderlos:
+        ges_beitr_midi_job_m += (
```
```suggestion
        gesamtbeitrag_midi_job_m += (
```

hmgaudecker commented on gettsim/taxes/zu_verst_eink/eink.py:
```diff
@@ -190,65 +97,12 @@ def eink_rente_zu_verst(
-) -> FloatSeries:
+    eink_selbst: float,
+    eink_abhängig_beschäftigt: float,
+    vermiet_eink: float,
```
Did we decide to turn this around? I see no reason to change the name (but might have suggested it previously, of course), especially since the official thing is "Einkommen aus Vermietung und Verpachtung", if I am not mistaken.

ChristianZimpelmann replied:
I also don't remember.. changed it to `eink_vermiet` now

hmgaudecker replied:
Why not leave it? We decided on it that way for a reason...

ChristianZimpelmann replied:
Now I am confused. I thought you prefer `eink_vermiet`. I don't have a strong preference here.

hmgaudecker commented on gettsim/taxes/zu_verst_eink/freibeträge.py:
```diff
@@ -119,22 +121,22 @@ def eink_st_altersfreib(
 
     """
-    out = bruttolohn_m * 0
     agelimit = eink_st_abzüge_params["altersentlastungsbetrag_altersgrenze"]
```
```suggestion
    altersgrenze = eink_st_abzüge_params["altersentlastungsbetrag_altersgrenze"]
```
No need to translate.

hmgaudecker commented on gettsim/tests/test_tax_transfer.py:
```diff
@@ -14,7 +14,7 @@
 
 
-YEARS = [2019, 2020, 2021]
+YEARS = [2019]
```
Why?

ChristianZimpelmann replied:
The current test data only contains test cases for 2019. (No clue, why the code looped over 2019-2021 before)

hmgaudecker commented on gettsim/tests/test_unterhalt_vorschuss.py:
```diff
@@ -1,5 +1,3 @@
-import itertools
```
Filename needs updating.

hmgaudecker commented on gettsim/transfers/benefit_checks/vermögens_checks.py:
```diff
@@ -1,22 +1,14 @@
 def _kinderzuschl_nach_vermög_check_m_hh(
-    _kinderzuschl_vor_vermög_check_m_hh: FloatSeries,
-    vermögen_hh: FloatSeries,
+    kinderzuschl_vorläufig_m_hh: float,
```
I am pretty sure we explicitly changed that in the big renaming thing. Why back to `vorläufig`? It is ambiguous what is meant by that.

hmgaudecker reviewed (commented):
Looks great, thanks! Mostly minor comments, not sure about some renamings.

*Note to self: continue at `elterngeld.py`

hmgaudecker commented on gettsim/transfers/kindergeld.py:
```diff
@@ -1,9 +1,16 @@
+def kindergeld_m_bis_1996(kindergeld_basis_m: float) -> float:
     """Kindergeld calculation until 1996.
 
     Until 1996 individuals could claim child allowance and recieve child benefit.
```
```suggestion
    Until 1996 individuals could claim Kinderfreibetrag and receive Kindergeld at the same time.
```

hmgaudecker commented on gettsim/transfers/kindergeld.py:
```diff
@@ -40,104 +45,50 @@ def kindergeld_m_ab_1997(
-    """
-    return kindergeld_m.groupby(hh_id).sum()
-
+    if kinderfreib_günstiger_tu:
```
out = 0 if kinderfreib_günstiger_tu else kindergeld_basis_m

hmgaudecker commented on gettsim/transfers/kindergeld.py:
```diff
@@ -156,12 +107,10 @@ def kindergeld_anspruch_nach_stunden(
-    out = out | (
-        (kindergeld_params["höchstalter"]["ohne_bedingungen"] <= alter)
-        & (alter <= kindergeld_params["höchstalter"]["mit_bedingungen"])
+    out = (alter < kindergeld_params["höchstalter"]["ohne_bedingungen"]) | (
```
Why changed from `<=` to `<` ?, also line below ?  Former seems to make more sense. Pls always test these edge cases!

hmgaudecker replied:
> Why do you think the former makes more sense?

Because I would interpret "höchstalter" as "last age still eligible", not "first age ineligible". Given what you write, we should change to 24 in yaml then.

ChristianZimpelmann replied:
> Because I would interpret "höchstalter" as "last age still eligible", not "first age ineligible".

Makes sense.

On the other hand, the numbers mentioned in the law are 18 and 25. I tend towards leaving the numbers as is and improve the parameter description.

Alternative would be to change numbers to 17 and 24 and adjust code accordingly.

hmgaudecker commented on gettsim/transfers/kindergeld.py:
```diff
@@ -196,57 +142,10 @@ def kindergeld_anspruch_nach_lohn(
-    out = out | (
-        (kindergeld_params["höchstalter"]["ohne_bedingungen"] <= alter)
-        & (alter <= kindergeld_params["höchstalter"]["mit_bedingungen"])
+    out = (alter < kindergeld_params["höchstalter"]["ohne_bedingungen"]) | (
```
Same as before. Why suddently a strict equals ?

hmgaudecker commented on gettsim/transfers/kinderzuschl/kinderzuschl.py:
```diff
@@ -49,97 +44,89 @@ def kinderzuschl_m_hh(
-    out = _kinderzuschl_nach_vermög_check_m_hh.copy()
-    cond = (~kinderzuschl_vorrang_hh & ~wohngeld_kinderzuschl_vorrang_hh) | (
+    if ((not kinderzuschl_vorrang_hh) & (not wohngeld_kinderzuschl_vorrang_hh)) | (
         anz_rentner_hh > 0
```
Is `anz_rentner_hh > 0` in the law or some simplification? If the latter, it might be useful to make a note for that point in time when we'll try to disentangle `grunds_im_alter` and `arbeitsl_geld_2` better.

ChristianZimpelmann replied:
I think the latter.. I have made a note

hmgaudecker commented on gettsim/transfers/kinderzuschl/kinderzuschl.py:
```diff
@@ -49,97 +44,89 @@ def kinderzuschl_m_hh(
+    kinderzuschl_kindereink_abzug_m_hh: float,
+    kinderzuschl_eink_anrechn_m_hh: float,
+) -> float:
+    """Calculate preliminary child benefit since 2005 until 06/2019.
```
```suggestion
    """Calculate Kinderzuschlag since 2005 until 06/2019. Whether Kinderzuschlag or Arbeitslosengeld 2 applies will be checked later.
```

hmgaudecker commented on gettsim/transfers/kinderzuschl/kinderzuschl.py:
```diff
@@ -49,97 +44,89 @@ def kinderzuschl_m_hh(
+    kinderzuschl_kindereink_abzug_m_hh: float,
+    kinderzuschl_eink_anrechn_m_hh: float,
+) -> float:
+    """Calculate preliminary child benefit since 07/2019.
```
```suggestion
    """Calculate Kinderzuschlag since 07/2019. Whether Kinderzuschlag or Arbeitslosengeld 2 applies will be checked later.
```

hmgaudecker commented on gettsim/transfers/kinderzuschl/kost_unterk.py:
```diff
@@ -103,56 +84,25 @@ def _kinderzuschl_wohnbedarf_eltern_anteil(
-    erwachsene_in_tu = tu_id.replace(anz_erwachsene_tu)
-    conditions = []
-    choices = []
     ex_min = kinderzuschl_params["exmin"]
```
```suggestion
    ex_min = kinderzuschl_params["existenzminimum"]
```

No need to abbreviate (and use two different abbreviations)

hmgaudecker commented on gettsim/transfers/rente.py:
```diff
@@ -67,20 +40,22 @@ def ges_rente_nach_grundr_m(
-
     # Return 0 if person not yet retired
-    out.loc[~rentner] = 0
+    if rentner:
```
one-liner

hmgaudecker commented on gettsim/transfers/rente.py:
```diff
@@ -109,14 +84,16 @@ def ges_rente_vor_grundr_m(
 
-    out = entgeltp_update * ges_rente_zugangsfaktor * rentenwert
+    # Return 0 if person not yet retired
+    if rentner:
```
one-lineer

hmgaudecker commented on gettsim/transfers/rente.py:
```diff
@@ -130,16 +107,17 @@ def rentenwert(wohnort_ost: BoolSeries, ges_rente_params: dict) -> FloatSeries:
-    return out
+    params = ges_rente_params["rentenwert"]
+
+    if wohnort_ost:
```
one-liner

hmgaudecker commented on gettsim/transfers/rente.py:
```diff
@@ -153,18 +131,17 @@ def rentenwert_vorjahr(wohnort_ost: BoolSeries, ges_rente_params: dict) -> Float
-    return out
+    params = ges_rente_params["rentenwert_vorjahr"]
+
+    if wohnort_ost:
```
one-liner

hmgaudecker commented on gettsim/transfers/rente.py:
```diff
@@ -218,32 +196,35 @@ def entgeltp_update_lohn(
-        bruttolohn_scaled_east.loc[wohnort_ost]
-        * ges_rente_params["umrechnung_entgeltp_beitrittsgebiet"]
-    )
+    if wohnort_ost:
```
Oh, maybe make a note/issue: I thought this parameters only applies to what was earned during GDR times, now current stuff ?

hmgaudecker commented on gettsim/transfers/wohngeld.py:
```diff
@@ -365,54 +331,49 @@ def wohngeld_miete_m_bis_2008(
+    constr_year = list(params_max_miete[1])[selected_bin_index]
+
+    # Calc maximal considered rent
+    max_definierte_hh_größe = max(i for i in params_max_miete if isinstance(i, int))
```
Please make note / issue to do more pre-processing of yaml contents. We don't want to repeat this a gazillion times, also not very robust.

hmgaudecker reviewed (commented):
Great! Just a couple of simplifications, notes, very few substantive points. Thanks!

ChristianZimpelmann commented on .github/workflows/continuous-integration-workflow.yaml:
```diff
@@ -16,7 +16,7 @@ jobs:
       matrix:
         os: ["ubuntu-latest", "macos-latest", "windows-latest"]
-        python-version: ["3.7", "3.8", "3.9"]
+        python-version: ["3.8", "3.9"]
```
Good point. Will include 3.10, too.

ChristianZimpelmann commented on environment_linux_jax.yml:
```diff
@@ -0,0 +1,31 @@
+name: gettsim
```
For the aggregation tests, it was still relevant. But I agree that we don't need them at the moment.

hmgaudecker replied:
Your reply does not seem to fit the comment here. Or am I missing something?

ChristianZimpelmann replied:
Yes! Copied it to the correct comment

ChristianZimpelmann commented on .github/workflows/continuous-integration-workflow.yaml:
```diff
@@ -29,17 +29,34 @@ jobs:
           pre-commit install -f --install-hooks
           pre-commit run --all-files
+
+  run-jax-tests:
```
Should belong here: For the aggregation tests, it was still relevant. But I agree that we don't need them at the moment.

hmgaudecker commented:
Also still needed: Search for "Series" (uppercase) and replace appropriately at least in the GETTSIM functions (not necessary to go in detail through `interface.py` right now). 

There are quite some instances in the docs that were not using "FloatSeries" and friends before.

ChristianZimpelmann commented on gettsim/aggregation_numpy.py:
```diff
@@ -0,0 +1,147 @@
+def grouped_max(column, group_id):
+    fail_if_dtype_of_group_id_not_int(group_id, agg_func="max")
+    fail_if_dtype_not_numeric_or_datetime(column, agg_func="max")
+    fill_value = (
```
I added the following:
```
    # numpy_groupies expects a `fill_value` which is assigned to groups which are not
    # present in the data (e.g. if the data only contains the group_id 0 and 4, group
    # 1-3 would get this value). This isn't relevant here because grouped_max returns
    # the values on the individual level. Nevertheless, we need to make sure that the
    # fill_value is convertible to the dtype of `column`.

ChristianZimpelmann commented on gettsim/demographic_vars.py:
```diff
@@ -27,301 +41,154 @@ def anz_kinder_bis_17_hh(
-    geburtsjahr: IntSeries, geburtsmonat: IntSeries, geburtstag: IntSeries
-) -> DateTimeSeries:
+def geburtstermin(
+    geburtsjahr: int, geburtsmonat: int, geburtstag: int
```
That would be a good change. I will open an issue. Let's stick with `geburtstag` until then.

ChristianZimpelmann commented on gettsim/transfers/elterngeld.py:
```diff
@@ -1,56 +1,18 @@
-    elterngeld_mehrlinge_bonus_m: FloatSeries,
+    elterngeld_eink_relev_m: float,
+    elternzeit_anspruch: bool,
+    elterngeld_eink_erlass_m: float,
```
also no idea. 

Would leave it as is for now.

ChristianZimpelmann commented on gettsim/transfers/grunds_im_alter.py:
```diff
@@ -158,20 +138,20 @@ def grunds_im_alter_eink_m(
+    )
 
-    return out
+    return max(out, 0.0)
```
Why remove .0? 

Vectorization leads to problems if some results are float and others int.

hmgaudecker replied:
Just readability / consistency. I did not realise that.

hmgaudecker commented:
On the phone and cannot directly reply at the right spots:

- I'd stick to vermietung and also vor_vermög_check (or the like) — simply
not reason to rename in this PR. Sorry, I see why that was confusing.
-17/24 + höchstalter, unless you come up with a really good name (I did
not, but that does not mean anything)

Thanks!

ChristianZimpelmann commented:
> I'd stick to vermietung and also vor_vermög_check (or the like) — simply
not reason to rename in this PR. Sorry, I see why that was confusing.

I am still not sure what your prefered name is. We have two variables:
- input variable on monthly basis: current version and version before this PR: `eink_vermiet_m`
- defined function which calculates the value on yearly basis: before this PR `eink_vermietung`, current version: `eink_vermiet`

Just let me know how you would like to change it.

-17/24 + höchstalter, unless you come up with a really good name (I did
not, but that does not mean anything)

How about `altersgrenze`? I still think we should, if possible, use the parameter values mentioned in the law.

- Otherwise, all comments should be addressed now.
- I haven't implemented the user-provided aggregation-dictionary, yet. Should we postpone?

hmgaudecker commented:
> > I'd stick to vermietung and also vor_vermög_check (or the like) — simply
> > not reason to rename in this PR. Sorry, I see why that was confusing.
> 
> I am still not sure what your prefered name is. We have two variables:
> 
>     input variable on monthly basis: current version and version before this PR: `eink_vermiet_m`
> 
>     defined function which calculates the value on yearly basis: before this PR `eink_vermietung`, current version: `eink_vermiet`

Ahh, I did not remember the former. Then we forgot to rename that, I believe. `eink_vermietung_m` / `eink_vermietung`, if that is not too much of a hassle

> How about `altersgrenze`? I still think we should, if possible, use the parameter values mentioned in the law.

Very good.

>     Otherwise, all comments should be addressed now.

Perfect! 
> 
>     I haven't implemented the user-provided aggregation-dictionary, yet. Should we postpone?

Sure, just open an issue. Or a PR directly, should be just a couple of hours' work, right? Then would be nice to have in release.

>     I tried to add python 3.10, but didn't get to run a working conda environment, yet. So let's postpone this for a bit.

Sure.
